File: camera.py
import numpy as np
from pyrr import Vector3

class Camera:
    def __init__(self, position, sensitivity=0.002):
        self.position = Vector3(position)
        self.yaw = 0.0
        self.pitch = 0.0
        self.sensitivity = sensitivity
        self.first_mouse = True
        self.lastX = 0
        self.lastY = 0

    def process_mouse(self, xpos, ypos):
        if self.first_mouse:
            self.lastX, self.lastY = xpos, ypos
            self.first_mouse = False
            return

        xoffset = xpos - self.lastX
        yoffset = self.lastY - ypos
        self.lastX, self.lastY = xpos, ypos

        self.yaw -= xoffset * self.sensitivity
        self.pitch += yoffset * self.sensitivity

        max_pitch = np.pi / 2.0 - 0.01
        self.pitch = np.clip(self.pitch, -max_pitch, max_pitch)

    def get_rotation_matrix(self):
        front = Vector3([
            np.cos(self.pitch) * np.sin(self.yaw),
            np.sin(self.pitch),
            np.cos(self.pitch) * np.cos(self.yaw)
        ]).normalized
        world_up = Vector3([0.0, 1.0, 0.0])
        right = front.cross(world_up).normalized
        up = right.cross(front).normalized
        return np.array([list(right), list(up), list(front)], dtype='f4'), front, right

File: src/camera.cpp
#include "camera.h"

Camera::Camera(glm::vec3 position, glm::vec3 up, float yaw, float pitch)
    : Front(glm::vec3(0.0f, 0.0f, -1.0f)), MovementSpeed(SPEED), 
      MouseSensitivity(SENSITIVITY), Fov(FOV) {
    Position = position;
    WorldUp = up;
    Yaw = yaw;
    Pitch = pitch;
    updateCameraVectors();
}

glm::mat4 Camera::getViewMatrix() {
    return glm::lookAt(Position, Position + Front, Up);
}

void Camera::processKeyboard(CameraMovement direction, float deltaTime) {
    float velocity = MovementSpeed * deltaTime;
    
    if (direction == FORWARD)
        Position += Front * velocity;
    if (direction == BACKWARD)
        Position -= Front * velocity;
    if (direction == LEFT)
        Position -= Right * velocity;
    if (direction == RIGHT)
        Position += Right * velocity;
    if (direction == UP)
        Position += Up * velocity;
    if (direction == DOWN)
        Position -= Up * velocity;
}

void Camera::processMouseMovement(float xoffset, float yoffset, GLboolean constrainPitch) {
    xoffset *= MouseSensitivity;
    yoffset *= MouseSensitivity;
    
    Yaw += xoffset;
    Pitch += yoffset;
    
    if (constrainPitch) {
        if (Pitch > 89.0f)
            Pitch = 89.0f;
        if (Pitch < -89.0f)
            Pitch = -89.0f;
    }
    
    updateCameraVectors();
}

void Camera::processMouseScroll(float yoffset) {
    const float scalingFactor = 1.1f;

    if (yoffset > 0) {
        MovementSpeed *= scalingFactor;
    } else if (yoffset < 0) {
        MovementSpeed /= scalingFactor;
    }
    const float minSpeed = 0.0000001f;
    const float maxSpeed = 50.0f;

    if (MovementSpeed < minSpeed)
        MovementSpeed = minSpeed;
    
    if (MovementSpeed > maxSpeed)
        MovementSpeed = maxSpeed;
}

void Camera::updateCameraVectors() {
    glm::vec3 front;
    front.x = cos(glm::radians(Yaw)) * cos(glm::radians(Pitch));
    front.y = sin(glm::radians(Pitch));
    front.z = sin(glm::radians(Yaw)) * cos(glm::radians(Pitch));
    Front = glm::normalize(front);
    
    Right = glm::normalize(glm::cross(Front, WorldUp));
    Up = glm::normalize(glm::cross(Right, Front));
}